allure-grpc/src/main/java/io/qameta/allure/grpc/AllureGrpc.java

```diff
@@ -0,0 +1,162 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Allure interceptor logger for gRPC.
+ * @author dtuchs (Dmitry Tuchs).
+ */
+public class AllureGrpc implements ClientInterceptor {
+
+    private static final Logger log = LoggerFactory.getLogger(AllureGrpc.class);
+    private static final JsonFormat.Printer jsonPrinter = JsonFormat.printer();
+
+    private String requestTemplatePath = "grpc-request.ftl";
+    private String responseTemplatePath = "grpc-response.ftl";
+
+    public AllureGrpc setRequestTemplate(final String templatePath) {
+        this.requestTemplatePath = templatePath;
+        return this;
+    }
+
+    public AllureGrpc setResponseTemplate(final String templatePath) {
+        this.responseTemplatePath = templatePath;
+        return this;
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+        final AttachmentProcessor<AttachmentData> processor = new DefaultAttachmentProcessor();
+
+        return new ForwardingClientCall.SimpleForwardingClientCall<>(
+                next.newCall(method, callOptions.withoutWaitForReady())) {
+
+            private String stepUuid;
+            private List<String> parsedResponses = new ArrayList<>();
+
+            @Override
+            public void sendMessage(ReqT message) {
+                stepUuid = UUID.randomUUID().toString();
+                Allure.getLifecycle().startStep(stepUuid, (new StepResult()).setName(
+                        "Send gRPC request to "
+                                + next.authority()
+                                + trimGrpcMethodName(method.getFullMethodName())
+                ));
+                try {
+                    GrpcRequestAttachment rpcRequestAttach = GrpcRequestAttachment.Builder
+                            .create("gRPC request", method.getFullMethodName())
+                            .setBody(jsonPrinter.print((MessageOrBuilder) message))
+                            .build();
+                    processor.addAttachment(rpcRequestAttach, new FreemarkerAttachmentRenderer(requestTemplatePath));
+                    super.sendMessage(message);
+                } catch (InvalidProtocolBufferException e) {
+                    log.warn("Can`t parse gRPC request", e);
+                } catch (Throwable e) {
+                    Allure.getLifecycle().updateStep((s) ->
+                            s.setStatus(ResultsUtils.getStatus(e).orElse(Status.BROKEN))
+                                    .setStatusDetails(ResultsUtils.getStatusDetails(e).orElse(null))
+                    );
+                    Allure.getLifecycle().stopStep(stepUuid);
+                    stepUuid = null;
+                }
+            }
+
+            @Override
+            public void start(Listener<RespT> responseListener, Metadata headers) {
```

Comment:
Could I suggest to attach headers and trailers as well? This can be very useful in cases where the application logic takes metadata into workflow.

Reply:
Added an optional feature - interceptResponseMetadata flag.
I think that in most cases there is no need to save metadata, but why not :)